The job **`x86_64-gnu-llvm-15`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/6326890956/job/17181474266) [(plain)](https://github.com/rust-lang/rust/commit/b97d311edf20fd462fea6d332db2bd89f8f3fbf5/checks/17181474266/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ACTION=__run_7
GITHUB_ACTIONS=true
GITHUB_ACTION_REF=
GITHUB_ACTION_REPOSITORY=
GITHUB_ACTOR=Urgau
GITHUB_API_URL=https://api.github.com
GITHUB_BASE_REF=master
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_02a45bdf-3248-478b-9a48-7eb863d18ea3
GITHUB_EVENT_NAME=pull_request
---
GITHUB_SERVER_URL=https://github.com
GITHUB_SHA=a0f5b0884709a7bf9b28c4259d65d81323795336
GITHUB_STATE=/home/runner/work/_temp/_runner_file_commands/save_state_02a45bdf-3248-478b-9a48-7eb863d18ea3
GITHUB_STEP_SUMMARY=/home/runner/work/_temp/_runner_file_commands/step_summary_02a45bdf-3248-478b-9a48-7eb863d18ea3
GITHUB_TRIGGERING_ACTOR=Urgau
GITHUB_WORKFLOW_REF=rust-lang/rust/.github/workflows/ci.yml@refs/pull/116199/merge
GITHUB_WORKFLOW_SHA=a0f5b0884709a7bf9b28c4259d65d81323795336
GITHUB_WORKSPACE=/home/runner/work/rust/rust
GOROOT_1_18_X64=/opt/hostedtoolcache/go/1.18.10/x64
---
status: exit status: 101
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/consts/const-vec-of-fns.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-vec-of-fns/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/consts/const-vec-of-fns/auxiliary"
stdout: none
--- stderr -------------------------------
##[error]error: internal compiler error: compiler/rustc_middle/src/ty/typeck_results.rs:494:9: node HirId(DefId(0:9 ~ const_vec_of_fns[c246]::closures).7) (expr `&mut [S(f as fn()), S(f as fn())]`) cannot be placed in TypeckResults with hir_owner DefId(0:10 ~ const_vec_of_fns[c246]::main)
thread 'rustc' panicked at /checkout/compiler/rustc_errors/src/lib.rs:1651:9:
Box<dyn Any>
stack backtrace:
   0:     0x7efc35e8acfc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc1627d5f0b96d1e5
   0:     0x7efc35e8acfc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc1627d5f0b96d1e5
   1:     0x7efc35ef8847 - core::fmt::write::h7e4f80807e4491d9
Build completed unsuccessfully in 0:12:21
   2:     0x7efc35e7eed5 - std::io::Write::write_fmt::hb2d0f452212aee30
   3:     0x7efc35e8aae5 - std::sys_common::backtrace::print::hba26cdc40ba151c5
   4:     0x7efc35e8dd3a - std::panicking::default_hook::{{closure}}::h4b34a8d808149f9a
   5:     0x7efc35e8da2e - std::panicking::default_hook::h10bd3fadd73b53b5
   6:     0x7efc36ae57c8 - <alloc[129812d629d9ff55]::boxed::Box<rustc_driver_impl[ab5ca913d2dc0d3a]::install_ice_hook::{closure#0}> as core[2cd07bc95e06c344]::ops::function::Fn<(&dyn for<'a, 'b> core[2cd07bc95e06c344]::ops::function::Fn<(&'a core[2cd07bc95e06c344]::panic::panic_info::PanicInfo<'b>,), Output = ()> + core[2cd07bc95e06c344]::marker::Sync + core[2cd07bc95e06c344]::marker::Send, &core[2cd07bc95e06c344]::panic::panic_info::PanicInfo)>>::call
   7:     0x7efc35e8e4a5 - std::panicking::rust_panic_with_hook::hbd27181e424b5422
   8:     0x7efc39598116 - std[d02fda7d125d4b97]::panicking::begin_panic::<rustc_errors[12cefd157dfa59cb]::ExplicitBug>::{closure#0}
   9:     0x7efc395924f6 - std[d02fda7d125d4b97]::sys_common::backtrace::__rust_end_short_backtrace::<std[d02fda7d125d4b97]::panicking::begin_panic<rustc_errors[12cefd157dfa59cb]::ExplicitBug>::{closure#0}, !>
  10:     0x7efc36a1a136 - std[d02fda7d125d4b97]::panicking::begin_panic::<rustc_errors[12cefd157dfa59cb]::ExplicitBug>
  11:     0x7efc394e13e7 - <rustc_errors[12cefd157dfa59cb]::HandlerInner>::bug::<alloc[129812d629d9ff55]::string::String>
  12:     0x7efc394e1279 - <rustc_errors[12cefd157dfa59cb]::Handler>::bug::<alloc[129812d629d9ff55]::string::String>
  13:     0x7efc3962a217 - rustc_middle[dab9cfca96648990]::util::bug::opt_span_bug_fmt::<rustc_span[4c5d015ad8bb46ac]::span_encoding::Span>::{closure#0}
  14:     0x7efc3961ff6c - rustc_middle[dab9cfca96648990]::ty::context::tls::with_opt::<rustc_middle[dab9cfca96648990]::util::bug::opt_span_bug_fmt<rustc_span[4c5d015ad8bb46ac]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7efc3961ff04 - rustc_middle[dab9cfca96648990]::ty::context::tls::with_context_opt::<rustc_middle[dab9cfca96648990]::ty::context::tls::with_opt<rustc_middle[dab9cfca96648990]::util::bug::opt_span_bug_fmt<rustc_span[4c5d015ad8bb46ac]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7efc36a1cca2 - rustc_middle[dab9cfca96648990]::util::bug::bug_fmt
  17:     0x7efc36a1e6aa - rustc_middle[dab9cfca96648990]::ty::typeck_results::invalid_hir_id_for_typeck_results
  18:     0x7efc39672aa4 - <rustc_middle[dab9cfca96648990]::ty::typeck_results::TypeckResults>::node_type_opt
  19:     0x7efc39672742 - <rustc_middle[dab9cfca96648990]::ty::typeck_results::TypeckResults>::node_type
  20:     0x7efc38eb8fe4 - <rustc_lint[248f027a0354a5d7]::reference_casting::InvalidReferenceCasting as rustc_lint[248f027a0354a5d7]::passes::LateLintPass>::check_expr
  21:     0x7efc38fada05 - <rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass as rustc_lint[248f027a0354a5d7]::passes::LateLintPass>::check_expr
  22:     0x7efc38ed4425 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  23:     0x7efc38f31f0f - rustc_hir[79ab3520dafcee8e]::intravisit::walk_expr::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  24:     0x7efc38ed4430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  25:     0x7efc38f32021 - rustc_hir[79ab3520dafcee8e]::intravisit::walk_expr::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  26:     0x7efc38ed4430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  27:     0x7efc38ed4430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  28:     0x7efc38ed4430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  29:     0x7efc38ed4430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  30:     0x7efc38ed26dc - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_nested_body
  31:     0x7efc38ed5223 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_fn
  32:     0x7efc38f33cb0 - rustc_hir[79ab3520dafcee8e]::intravisit::walk_item::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  33:     0x7efc38ecde9e - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_nested_item
  34:     0x7efc38f3016c - rustc_hir[79ab3520dafcee8e]::intravisit::walk_mod::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  35:     0x7efc38ed8b24 - rustc_lint[248f027a0354a5d7]::late::late_lint_mod::<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>
  36:     0x7efc38fa9c7d - rustc_lint[248f027a0354a5d7]::lint_mod
  37:     0x7efc3838bc37 - rustc_query_impl[7f9ca928777a02b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f9ca928777a02b3]::query_impl::lint_mod::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 0usize]>>
  38:     0x7efc38510609 - <rustc_query_impl[7f9ca928777a02b3]::query_impl::lint_mod::dynamic_query::{closure#2} as core[2cd07bc95e06c344]::ops::function::FnOnce<(rustc_middle[dab9cfca96648990]::ty::context::TyCtxt, rustc_span[4c5d015ad8bb46ac]::def_id::LocalModDefId)>>::call_once
  39:     0x7efc3865761f - rustc_query_system[43d01bf7f334935f]::query::plumbing::try_execute_query::<rustc_query_impl[7f9ca928777a02b3]::DynamicConfig<rustc_query_system[43d01bf7f334935f]::query::caches::DefaultCache<rustc_span[4c5d015ad8bb46ac]::def_id::LocalModDefId, rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[7f9ca928777a02b3]::plumbing::QueryCtxt, false>
  40:     0x7efc385418cc - rustc_query_impl[7f9ca928777a02b3]::query_impl::lint_mod::get_query_non_incr::__rust_end_short_backtrace
  41:     0x7efc38ecb25e - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures[4694bdc52603536e]::sync::parallel::disabled::par_for_each_in<&[rustc_hir[79ab3520dafcee8e]::hir_id::OwnerId], <rustc_middle[dab9cfca96648990]::hir::map::Map>::par_for_each_module<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  42:     0x7efc38f397cb - rustc_data_structures[4694bdc52603536e]::sync::parallel::disabled::par_for_each_in::<&[rustc_hir[79ab3520dafcee8e]::hir_id::OwnerId], <rustc_middle[dab9cfca96648990]::hir::map::Map>::par_for_each_module<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}::{closure#0}::{closure#0}>::{closure#0}>
  43:     0x7efc38ecbedd - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  44:     0x7efc38f3995d - rustc_data_structures[4694bdc52603536e]::sync::parallel::disabled::join::<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#0}, rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}, (), ()>
  45:     0x7efc38ed8ee0 - rustc_lint[248f027a0354a5d7]::late::check_crate
  46:     0x7efc36d0cb39 - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}::{closure#0}::{closure#2}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  47:     0x7efc36c5511c - <rustc_data_structures[4694bdc52603536e]::sync::parallel::ParallelGuard>::run::<(), rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}::{closure#0}::{closure#2}>
  48:     0x7efc36d0d4cc - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  49:     0x7efc36c5581c - <rustc_data_structures[4694bdc52603536e]::sync::parallel::ParallelGuard>::run::<(), rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}>
  50:     0x7efc36d038e7 - <rustc_session[ac082b45030d6a13]::session::Session>::time::<(), rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}>
  51:     0x7efc36c32dab - rustc_interface[a919bb2278be0b34]::passes::analysis
  52:     0x7efc3838b475 - rustc_query_impl[7f9ca928777a02b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f9ca928777a02b3]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 1usize]>>
  53:     0x7efc38495475 - <rustc_query_impl[7f9ca928777a02b3]::query_impl::analysis::dynamic_query::{closure#2} as core[2cd07bc95e06c344]::ops::function::FnOnce<(rustc_middle[dab9cfca96648990]::ty::context::TyCtxt, ())>>::call_once
  54:     0x7efc3862a2bd - rustc_query_system[43d01bf7f334935f]::query::plumbing::try_execute_query::<rustc_query_impl[7f9ca928777a02b3]::DynamicConfig<rustc_query_system[43d01bf7f334935f]::query::caches::SingleCache<rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[7f9ca928777a02b3]::plumbing::QueryCtxt, false>
  55:     0x7efc384d5a7d - rustc_query_impl[7f9ca928777a02b3]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  56:     0x7efc36abed05 - <rustc_middle[dab9cfca96648990]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}::{closure#2}::{closure#6}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>
  57:     0x7efc36ae84a4 - <scoped_tls[898d3446c976b06b]::ScopedKey<rustc_span[4c5d015ad8bb46ac]::SessionGlobals>>::set::<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>
  58:     0x7efc36ae4327 - std[d02fda7d125d4b97]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[a919bb2278be0b34]::util::run_in_thread_with_globals<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>
  59:     0x7efc36ad5ffe - std[d02fda7d125d4b97]::panicking::try::<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<<std[d02fda7d125d4b97]::thread::Builder>::spawn_unchecked_<rustc_interface[a919bb2278be0b34]::util::run_in_thread_with_globals<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  60:     0x7efc36ae46f0 - <<std[d02fda7d125d4b97]::thread::Builder>::spawn_unchecked_<rustc_interface[a919bb2278be0b34]::util::run_in_thread_with_globals<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#1} as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  62:     0x7efc35c34b43 - <unknown>
  63:     0x7efc35cc6a00 - <unknown>
  64:                0x0 - <unknown>


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.74.0-nightly (a0f5b0884 2023-09-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [lint_mod] linting top-level module
end of query stack
error: aborting due to previous error
------------------------------------------

---
status: exit status: 101
command: RUSTC_ICE="0" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/lint-const-item-mutation.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-const-item-mutation" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-const-item-mutation/auxiliary"
stdout: none
--- stderr -------------------------------
warning: attempting to modify a `const` item
   |
   |
LL |     ARRAY[0] = 5; //~ WARN attempting to modify
   |
   |
   = note: each usage of a `const` item creates a new temporary; the original `const` item will not be modified
note: `const` item defined here
   |
   |
LL | const ARRAY: [u8; 1] = [25];
   = note: `#[warn(const_item_mutation)]` on by default


warning: attempting to modify a `const` item
   |
   |
LL |     MY_STRUCT.field = false; //~ WARN attempting to modify
   |
   |
   = note: each usage of a `const` item creates a new temporary; the original `const` item will not be modified
note: `const` item defined here
   |
   |
LL | const MY_STRUCT: MyStruct = MyStruct { field: true, inner_array: ['a'], raw_ptr: 2 as *mut u8 };


warning: attempting to modify a `const` item
   |
   |
LL |     MY_STRUCT.inner_array[0] = 'b'; //~ WARN attempting to modify
   |
   |
   = note: each usage of a `const` item creates a new temporary; the original `const` item will not be modified
note: `const` item defined here
   |
   |
LL | const MY_STRUCT: MyStruct = MyStruct { field: true, inner_array: ['a'], raw_ptr: 2 as *mut u8 };

warning: taking a mutable reference to a `const` item
##[warning]  --> /checkout/tests/ui/lint/lint-const-item-mutation.rs:40:5
   |
   |
LL |     MY_STRUCT.use_mut(); //~ WARN taking
   |
   |
   = note: each usage of a `const` item creates a new temporary
   = note: the mutable reference will refer to this temporary, not the original `const` item
note: mutable reference created due to call to this method
   |
LL |     fn use_mut(&mut self) {}
   |     ^^^^^^^^^^^^^^^^^^^^^
note: `const` item defined here
note: `const` item defined here
  --> /checkout/tests/ui/lint/lint-const-item-mutation.rs:27:1
   |
LL | const MY_STRUCT: MyStruct = MyStruct { field: true, inner_array: ['a'], raw_ptr: 2 as *mut u8 };

warning: taking a mutable reference to a `const` item
##[warning]  --> /checkout/tests/ui/lint/lint-const-item-mutation.rs:41:5
   |
   |
LL |     &mut MY_STRUCT; //~ WARN taking
   |
   |
   = note: each usage of a `const` item creates a new temporary
   = note: the mutable reference will refer to this temporary, not the original `const` item
note: `const` item defined here
   |
   |
LL | const MY_STRUCT: MyStruct = MyStruct { field: true, inner_array: ['a'], raw_ptr: 2 as *mut u8 };

warning: taking a mutable reference to a `const` item
##[warning]  --> /checkout/tests/ui/lint/lint-const-item-mutation.rs:42:5
   |
   |
LL |     (&mut MY_STRUCT).use_mut(); //~ WARN taking
   |
   |
   = note: each usage of a `const` item creates a new temporary
   = note: the mutable reference will refer to this temporary, not the original `const` item
note: `const` item defined here
   |
   |
LL | const MY_STRUCT: MyStruct = MyStruct { field: true, inner_array: ['a'], raw_ptr: 2 as *mut u8 };


warning: attempting to modify a `const` item
   |
   |
LL |     MUTABLE2.msg = "wow"; //~ WARN attempting to modify
   |
   |
   = note: each usage of a `const` item creates a new temporary; the original `const` item will not be modified
note: `const` item defined here
   |
   |
LL | const MUTABLE2: Mutable2 = Mutable2 { msg: "", other: String::new() };

warning: taking a mutable reference to a `const` item
##[warning]  --> /checkout/tests/ui/lint/lint-const-item-mutation.rs:55:5
   |
   |
LL |     VEC.push(0); //~ WARN taking a mutable reference to a `const` item
   |
   |
   = note: each usage of a `const` item creates a new temporary
   = note: the mutable reference will refer to this temporary, not the original `const` item
note: mutable reference created due to call to this method
  --> /rustc/FAKE_PREFIX/library/alloc/src/vec/mod.rs:1878:5
note: `const` item defined here
   |
   |
LL | const VEC: Vec<i32> = Vec::new();


##[error]error: internal compiler error: compiler/rustc_middle/src/ty/typeck_results.rs:494:9: node HirId(DefId(0:20 ~ lint_const_item_mutation[3a72]::RAW_PTR).4) (expr `1 as *mut u8`) cannot be placed in TypeckResults with hir_owner DefId(0:29 ~ lint_const_item_mutation[3a72]::main)
thread 'rustc' panicked at /checkout/compiler/rustc_errors/src/lib.rs:1651:9:
Box<dyn Any>
stack backtrace:
   0:     0x7fb3fd532cfc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc1627d5f0b96d1e5
   0:     0x7fb3fd532cfc - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hc1627d5f0b96d1e5
   1:     0x7fb3fd5a0847 - core::fmt::write::h7e4f80807e4491d9
   2:     0x7fb3fd526ed5 - std::io::Write::write_fmt::hb2d0f452212aee30
   3:     0x7fb3fd532ae5 - std::sys_common::backtrace::print::hba26cdc40ba151c5
   4:     0x7fb3fd535d3a - std::panicking::default_hook::{{closure}}::h4b34a8d808149f9a
   5:     0x7fb3fd535a2e - std::panicking::default_hook::h10bd3fadd73b53b5
   6:     0x7fb3fe18d7c8 - <alloc[129812d629d9ff55]::boxed::Box<rustc_driver_impl[ab5ca913d2dc0d3a]::install_ice_hook::{closure#0}> as core[2cd07bc95e06c344]::ops::function::Fn<(&dyn for<'a, 'b> core[2cd07bc95e06c344]::ops::function::Fn<(&'a core[2cd07bc95e06c344]::panic::panic_info::PanicInfo<'b>,), Output = ()> + core[2cd07bc95e06c344]::marker::Sync + core[2cd07bc95e06c344]::marker::Send, &core[2cd07bc95e06c344]::panic::panic_info::PanicInfo)>>::call
   7:     0x7fb3fd5364a5 - std::panicking::rust_panic_with_hook::hbd27181e424b5422
   8:     0x7fb400c40116 - std[d02fda7d125d4b97]::panicking::begin_panic::<rustc_errors[12cefd157dfa59cb]::ExplicitBug>::{closure#0}
   9:     0x7fb400c3a4f6 - std[d02fda7d125d4b97]::sys_common::backtrace::__rust_end_short_backtrace::<std[d02fda7d125d4b97]::panicking::begin_panic<rustc_errors[12cefd157dfa59cb]::ExplicitBug>::{closure#0}, !>
  10:     0x7fb3fe0c2136 - std[d02fda7d125d4b97]::panicking::begin_panic::<rustc_errors[12cefd157dfa59cb]::ExplicitBug>
  11:     0x7fb400b893e7 - <rustc_errors[12cefd157dfa59cb]::HandlerInner>::bug::<alloc[129812d629d9ff55]::string::String>
  12:     0x7fb400b89279 - <rustc_errors[12cefd157dfa59cb]::Handler>::bug::<alloc[129812d629d9ff55]::string::String>
  13:     0x7fb400cd2217 - rustc_middle[dab9cfca96648990]::util::bug::opt_span_bug_fmt::<rustc_span[4c5d015ad8bb46ac]::span_encoding::Span>::{closure#0}
  14:     0x7fb400cc7f6c - rustc_middle[dab9cfca96648990]::ty::context::tls::with_opt::<rustc_middle[dab9cfca96648990]::util::bug::opt_span_bug_fmt<rustc_span[4c5d015ad8bb46ac]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  15:     0x7fb400cc7f04 - rustc_middle[dab9cfca96648990]::ty::context::tls::with_context_opt::<rustc_middle[dab9cfca96648990]::ty::context::tls::with_opt<rustc_middle[dab9cfca96648990]::util::bug::opt_span_bug_fmt<rustc_span[4c5d015ad8bb46ac]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  16:     0x7fb3fe0c4ca2 - rustc_middle[dab9cfca96648990]::util::bug::bug_fmt
  17:     0x7fb3fe0c66aa - rustc_middle[dab9cfca96648990]::ty::typeck_results::invalid_hir_id_for_typeck_results
  18:     0x7fb400d1aaa4 - <rustc_middle[dab9cfca96648990]::ty::typeck_results::TypeckResults>::node_type_opt
  19:     0x7fb400d1a742 - <rustc_middle[dab9cfca96648990]::ty::typeck_results::TypeckResults>::node_type
  20:     0x7fb400560fe4 - <rustc_lint[248f027a0354a5d7]::reference_casting::InvalidReferenceCasting as rustc_lint[248f027a0354a5d7]::passes::LateLintPass>::check_expr
  21:     0x7fb400655a05 - <rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass as rustc_lint[248f027a0354a5d7]::passes::LateLintPass>::check_expr
  22:     0x7fb40057c425 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  23:     0x7fb40057cf4f - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_stmt
  24:     0x7fb4005da19c - rustc_hir[79ab3520dafcee8e]::intravisit::walk_expr::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  25:     0x7fb40057c430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  26:     0x7fb40057cf4f - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_stmt
  27:     0x7fb4005da19c - rustc_hir[79ab3520dafcee8e]::intravisit::walk_expr::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  28:     0x7fb40057c430 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_expr::{closure#0}
  29:     0x7fb40057a6dc - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_nested_body
  30:     0x7fb40057d223 - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_fn
  31:     0x7fb4005dbcb0 - rustc_hir[79ab3520dafcee8e]::intravisit::walk_item::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  32:     0x7fb400575e9e - <rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass> as rustc_hir[79ab3520dafcee8e]::intravisit::Visitor>::visit_nested_item
  33:     0x7fb4005d816c - rustc_hir[79ab3520dafcee8e]::intravisit::walk_mod::<rustc_lint[248f027a0354a5d7]::late::LateContextAndPass<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>>
  34:     0x7fb400580b24 - rustc_lint[248f027a0354a5d7]::late::late_lint_mod::<rustc_lint[248f027a0354a5d7]::BuiltinCombinedModuleLateLintPass>
  35:     0x7fb400651c7d - rustc_lint[248f027a0354a5d7]::lint_mod
  36:     0x7fb3ffa33c37 - rustc_query_impl[7f9ca928777a02b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f9ca928777a02b3]::query_impl::lint_mod::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 0usize]>>
  37:     0x7fb3ffbb8609 - <rustc_query_impl[7f9ca928777a02b3]::query_impl::lint_mod::dynamic_query::{closure#2} as core[2cd07bc95e06c344]::ops::function::FnOnce<(rustc_middle[dab9cfca96648990]::ty::context::TyCtxt, rustc_span[4c5d015ad8bb46ac]::def_id::LocalModDefId)>>::call_once
  38:     0x7fb3ffcff61f - rustc_query_system[43d01bf7f334935f]::query::plumbing::try_execute_query::<rustc_query_impl[7f9ca928777a02b3]::DynamicConfig<rustc_query_system[43d01bf7f334935f]::query::caches::DefaultCache<rustc_span[4c5d015ad8bb46ac]::def_id::LocalModDefId, rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 0usize]>>, false, false, false>, rustc_query_impl[7f9ca928777a02b3]::plumbing::QueryCtxt, false>
  39:     0x7fb3ffbe98cc - rustc_query_impl[7f9ca928777a02b3]::query_impl::lint_mod::get_query_non_incr::__rust_end_short_backtrace
  40:     0x7fb40057325e - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures[4694bdc52603536e]::sync::parallel::disabled::par_for_each_in<&[rustc_hir[79ab3520dafcee8e]::hir_id::OwnerId], <rustc_middle[dab9cfca96648990]::hir::map::Map>::par_for_each_module<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}::{closure#0}::{closure#0}>::{closure#0}>::{closure#0}::{closure#0}::{closure#0}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  41:     0x7fb4005e17cb - rustc_data_structures[4694bdc52603536e]::sync::parallel::disabled::par_for_each_in::<&[rustc_hir[79ab3520dafcee8e]::hir_id::OwnerId], <rustc_middle[dab9cfca96648990]::hir::map::Map>::par_for_each_module<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}::{closure#0}::{closure#0}>::{closure#0}>
  42:     0x7fb400573edd - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  43:     0x7fb4005e195d - rustc_data_structures[4694bdc52603536e]::sync::parallel::disabled::join::<rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#0}, rustc_lint[248f027a0354a5d7]::late::check_crate::{closure#1}, (), ()>
  44:     0x7fb400580ee0 - rustc_lint[248f027a0354a5d7]::late::check_crate
  45:     0x7fb3fe3b4b39 - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}::{closure#0}::{closure#2}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  46:     0x7fb3fe2fd11c - <rustc_data_structures[4694bdc52603536e]::sync::parallel::ParallelGuard>::run::<(), rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}::{closure#0}::{closure#2}>
  47:     0x7fb3fe3b54cc - <core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}> as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once
  48:     0x7fb3fe2fd81c - <rustc_data_structures[4694bdc52603536e]::sync::parallel::ParallelGuard>::run::<(), rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}::{closure#1}::{closure#0}>
  49:     0x7fb3fe3ab8e7 - <rustc_session[ac082b45030d6a13]::session::Session>::time::<(), rustc_interface[a919bb2278be0b34]::passes::analysis::{closure#6}>
  50:     0x7fb3fe2dadab - rustc_interface[a919bb2278be0b34]::passes::analysis
  51:     0x7fb3ffa33475 - rustc_query_impl[7f9ca928777a02b3]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[7f9ca928777a02b3]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 1usize]>>
  52:     0x7fb3ffb3d475 - <rustc_query_impl[7f9ca928777a02b3]::query_impl::analysis::dynamic_query::{closure#2} as core[2cd07bc95e06c344]::ops::function::FnOnce<(rustc_middle[dab9cfca96648990]::ty::context::TyCtxt, ())>>::call_once
  53:     0x7fb3ffcd22bd - rustc_query_system[43d01bf7f334935f]::query::plumbing::try_execute_query::<rustc_query_impl[7f9ca928777a02b3]::DynamicConfig<rustc_query_system[43d01bf7f334935f]::query::caches::SingleCache<rustc_middle[dab9cfca96648990]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[7f9ca928777a02b3]::plumbing::QueryCtxt, false>
  54:     0x7fb3ffb7da7d - rustc_query_impl[7f9ca928777a02b3]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  55:     0x7fb3fe166d05 - <rustc_middle[dab9cfca96648990]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}::{closure#2}::{closure#6}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>
  56:     0x7fb3fe1904a4 - <scoped_tls[898d3446c976b06b]::ScopedKey<rustc_span[4c5d015ad8bb46ac]::SessionGlobals>>::set::<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>
  57:     0x7fb3fe18c327 - std[d02fda7d125d4b97]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[a919bb2278be0b34]::util::run_in_thread_with_globals<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>
  58:     0x7fb3fe17dffe - std[d02fda7d125d4b97]::panicking::try::<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, core[2cd07bc95e06c344]::panic::unwind_safe::AssertUnwindSafe<<std[d02fda7d125d4b97]::thread::Builder>::spawn_unchecked_<rustc_interface[a919bb2278be0b34]::util::run_in_thread_with_globals<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  59:     0x7fb3fe18c6f0 - <<std[d02fda7d125d4b97]::thread::Builder>::spawn_unchecked_<rustc_interface[a919bb2278be0b34]::util::run_in_thread_with_globals<rustc_interface[a919bb2278be0b34]::interface::run_compiler<core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>, rustc_driver_impl[ab5ca913d2dc0d3a]::run_compiler::{closure#1}>::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[2cd07bc95e06c344]::result::Result<(), rustc_span[4c5d015ad8bb46ac]::ErrorGuaranteed>>::{closure#1} as core[2cd07bc95e06c344]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  61:     0x7fb3fd2dcb43 - <unknown>
  62:     0x7fb3fd36ea00 - <unknown>
  63:                0x0 - <unknown>


note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.74.0-nightly (a0f5b0884 2023-09-27) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [lint_mod] linting top-level module
end of query stack
error: aborting due to previous error; 8 warnings emitted
------------------------------------------

```

</details>

--- a/compiler/rustc_lint/src/context.rs
+++ b/compiler/rustc_lint/src/context.rs
@@ -1317,6 +1317,40 @@ impl<'tcx> LateContext<'tcx> {
     }
 
     /// If the given expression is a local binding, find the initializer expression.
+    /// If that initializer expression is another local binding, find its initializer again.
+    ///
+    /// This process repeats as long as possible (but usually no more than once).
+    /// Type-check adjustments are not taken in account in this function.
+    ///
+    /// Examples:
+    /// ```
+    /// let abc = 1;
+    /// let def = abc + 2;
+    /// //        ^^^^^^^ output
+    /// let def = def;
+    /// dbg!(def);
+    /// //   ^^^ input
+    /// ```
+    pub fn expr_or_init<'a>(&self, mut expr: &'a hir::Expr<'tcx>) -> &'a hir::Expr<'tcx> {
+        expr = expr.peel_blocks();
+
+        while let hir::ExprKind::Path(ref qpath) = expr.kind
+            && let Some(parent_node) = match self.qpath_res(qpath, expr.hir_id) {
+                Res::Local(hir_id) => self.tcx.hir().find_parent(hir_id),
+                _ => None,
+            }
+            && let Some(init) = match parent_node {
+                hir::Node::Expr(expr) => Some(expr),
+                hir::Node::Local(hir::Local { init, .. }) => *init,
+                _ => None
+            }
+        {
+            expr = init.peel_blocks();
+        }
+        expr
+    }
+
+    /// If the given expression is a local binding, find the initializer expression.
     /// If that initializer expression is another local or **outside** (`const`/`static`)
     /// binding, find its initializer again.
     ///
@@ -1338,7 +1372,10 @@ impl<'tcx> LateContext<'tcx> {
     /// dbg!(def);
     /// //   ^^^ input
     /// ```
-    pub fn expr_or_init<'a>(&self, mut expr: &'a hir::Expr<'tcx>) -> &'a hir::Expr<'tcx> {
+    pub fn expr_or_init_with_outside_body<'a>(
+        &self,
+        mut expr: &'a hir::Expr<'tcx>,
+    ) -> &'a hir::Expr<'tcx> {
         expr = expr.peel_blocks();
 
         while let hir::ExprKind::Path(ref qpath) = expr.kind

--- a/compiler/rustc_lint/src/invalid_from_utf8.rs
+++ b/compiler/rustc_lint/src/invalid_from_utf8.rs
@@ -84,9 +84,9 @@ impl<'tcx> LateLintPass<'tcx> for InvalidFromUtf8 {
                 )
             };
 
-            let mut init = cx.expr_or_init(arg);
+            let mut init = cx.expr_or_init_with_outside_body(arg);
             while let ExprKind::AddrOf(.., inner) = init.kind {
-                init = cx.expr_or_init(inner);
+                init = cx.expr_or_init_with_outside_body(inner);
             }
             match init.kind {
                 ExprKind::Lit(Spanned { node: lit, .. }) => {

--- a/compiler/rustc_lint/src/reference_casting.rs
+++ b/compiler/rustc_lint/src/reference_casting.rs
@@ -104,99 +104,51 @@ fn is_operation_we_care_about<'tcx>(
     deref_assign_or_addr_of(e).or_else(|| ptr_write(cx, e))
 }
 
-fn is_cast_from_const_to_mut<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'tcx>) -> bool {
-    let e = e.peel_blocks();
+fn is_cast_from_const_to_mut<'tcx>(cx: &LateContext<'tcx>, orig_expr: &'tcx Expr<'tcx>) -> bool {
+    let mut need_check_freeze = false;
+    let mut e = orig_expr;
 
-    fn from_casts<'tcx>(
-        cx: &LateContext<'tcx>,
-        e: &'tcx Expr<'tcx>,
-        need_check_freeze: &mut bool,
-    ) -> Option<&'tcx Expr<'tcx>> {
-        // <expr> as *mut ...
-        let mut e = if let ExprKind::Cast(e, t) = e.kind
-            && let ty::RawPtr(TypeAndMut { mutbl: Mutability::Mut, .. }) = cx.typeck_results().node_type(t.hir_id).kind() {
-            e
-        // <expr>.cast_mut()
+    let end_ty = cx.typeck_results().node_type(orig_expr.hir_id);
+
+    // Bail out early if the end type is **not** a mutable pointer.
+    if !matches!(end_ty.kind(), ty::RawPtr(TypeAndMut { ty: _, mutbl: Mutability::Mut })) {
+        return false;
+    }
+
+    loop {
+        e = e.peel_blocks();
+        // <expr> as ...
+        e = if let ExprKind::Cast(expr, _) = e.kind {
+            expr
+        // <expr>.cast(), <expr>.cast_mut() or <expr>.cast_const()
         } else if let ExprKind::MethodCall(_, expr, [], _) = e.kind
             && let Some(def_id) = cx.typeck_results().type_dependent_def_id(e.hir_id)
-            && cx.tcx.is_diagnostic_item(sym::ptr_cast_mut, def_id) {
+            && matches!(
+                cx.tcx.get_diagnostic_name(def_id),
+                Some(sym::ptr_cast | sym::const_ptr_cast | sym::ptr_cast_mut | sym::ptr_cast_const)
+            )
+        {
             expr
-        // UnsafeCell::raw_get(<expr>)
+        // ptr::from_ref(<expr>), UnsafeCell::raw_get(<expr>) or mem::transmute<_, _>(<expr>)
         } else if let ExprKind::Call(path, [arg]) = e.kind
             && let ExprKind::Path(ref qpath) = path.kind
             && let Some(def_id) = cx.qpath_res(qpath, path.hir_id).opt_def_id()
-            && cx.tcx.is_diagnostic_item(sym::unsafe_cell_raw_get, def_id)
+            && matches!(
+                cx.tcx.get_diagnostic_name(def_id),
+                Some(sym::ptr_from_ref | sym::unsafe_cell_raw_get | sym::transmute)
+            )
         {
-            *need_check_freeze = true;
+            if cx.tcx.is_diagnostic_item(sym::unsafe_cell_raw_get, def_id) {
+                need_check_freeze = true;
+            }
             arg
         } else {
-            return None;
+            break;
         };
-
-        let mut had_at_least_one_cast = false;
-        loop {
-            e = e.peel_blocks();
-            // <expr> as *mut/const ... or <expr> as <uint>
-            e = if let ExprKind::Cast(expr, t) = e.kind
-                && matches!(cx.typeck_results().node_type(t.hir_id).kind(), ty::RawPtr(_) | ty::Uint(_))  {
-                had_at_least_one_cast = true;
-                expr
-            // <expr>.cast(), <expr>.cast_mut() or <expr>.cast_const()
-            } else if let ExprKind::MethodCall(_, expr, [], _) = e.kind
-                && let Some(def_id) = cx.typeck_results().type_dependent_def_id(e.hir_id)
-                && matches!(
-                    cx.tcx.get_diagnostic_name(def_id),
-                    Some(sym::ptr_cast | sym::const_ptr_cast | sym::ptr_cast_mut | sym::ptr_cast_const)
-                )
-            {
-                had_at_least_one_cast = true;
-                expr
-            // ptr::from_ref(<expr>) or UnsafeCell::raw_get(<expr>)
-            } else if let ExprKind::Call(path, [arg]) = e.kind
-                && let ExprKind::Path(ref qpath) = path.kind
-                && let Some(def_id) = cx.qpath_res(qpath, path.hir_id).opt_def_id()
-                && matches!(
-                    cx.tcx.get_diagnostic_name(def_id),
-                    Some(sym::ptr_from_ref | sym::unsafe_cell_raw_get)
-                )
-            {
-                if cx.tcx.is_diagnostic_item(sym::unsafe_cell_raw_get, def_id) {
-                    *need_check_freeze = true;
-                }
-                return Some(arg);
-            } else if had_at_least_one_cast {
-                return Some(e);
-            } else {
-                return None;
-            };
-        }
     }
 
-    fn from_transmute<'tcx>(
-        cx: &LateContext<'tcx>,
-        e: &'tcx Expr<'tcx>,
-    ) -> Option<&'tcx Expr<'tcx>> {
-        // mem::transmute::<_, *mut _>(<expr>)
-        if let ExprKind::Call(path, [arg]) = e.kind
-            && let ExprKind::Path(ref qpath) = path.kind
-            && let Some(def_id) = cx.qpath_res(qpath, path.hir_id).opt_def_id()
-            && cx.tcx.is_diagnostic_item(sym::transmute, def_id)
-            && let ty::RawPtr(TypeAndMut { mutbl: Mutability::Mut, .. }) = cx.typeck_results().node_type(e.hir_id).kind() {
-            Some(arg)
-        } else {
-            None
-        }
-    }
-
-    let mut need_check_freeze = false;
-    let Some(e) = from_casts(cx, e, &mut need_check_freeze).or_else(|| from_transmute(cx, e))
-    else {
-        return false;
-    };
-
-    let e = e.peel_blocks();
-    let node_type = cx.typeck_results().node_type(e.hir_id);
-    if let ty::Ref(_, inner_ty, Mutability::Not) = node_type.kind() {
+    let start_ty = cx.typeck_results().node_type(e.hir_id);
+    if let ty::Ref(_, inner_ty, Mutability::Not) = start_ty.kind() {
         // If an UnsafeCell method is involved we need to additionaly check the
         // inner type for the presence of the Freeze trait (ie does NOT contain
         // an UnsafeCell), since in that case we would incorrectly lint on valid casts.

--- a/tests/ui/lint/reference_casting.rs
+++ b/tests/ui/lint/reference_casting.rs
@@ -36,6 +36,8 @@ unsafe fn ref_to_mut() {
     //~^ ERROR casting `&T` to `&mut T` is undefined behavior
     let _num = &mut *std::mem::transmute::<_, *mut i32>(num);
     //~^ ERROR casting `&T` to `&mut T` is undefined behavior
+    let _num = &mut *(std::mem::transmute::<_, *mut i32>(num) as *mut i32);
+    //~^ ERROR casting `&T` to `&mut T` is undefined behavior
     let _num = &mut *std::cell::UnsafeCell::raw_get(
     //~^ ERROR casting `&T` to `&mut T` is undefined behavior
         num as *const i32 as *const std::cell::UnsafeCell<i32>
@@ -51,6 +53,8 @@ unsafe fn ref_to_mut() {
     let _num = &mut *deferred_rebind;
     //~^ ERROR casting `&T` to `&mut T` is undefined behavior
     let _num = &mut *(num as *const _ as usize as *mut i32);
+    //~^ ERROR casting `&T` to `&mut T` is undefined behavior
+    let _num = &mut *(std::mem::transmute::<_, *mut _>(num as *const i32) as *mut i32);
     //~^ ERROR casting `&T` to `&mut T` is undefined behavior
 
     static NUM: &'static i32 = &2;
@@ -94,6 +98,8 @@ unsafe fn assign_to_ref() {
     *(std::ptr::from_ref({ num }) as *mut i32) += 1;
     //~^ ERROR assigning to `&T` is undefined behavior
     *std::mem::transmute::<_, *mut i32>(num) += 1;
+    //~^ ERROR assigning to `&T` is undefined behavior
+    *(std::mem::transmute::<_, *mut i32>(num) as *mut i32) += 1;
     //~^ ERROR assigning to `&T` is undefined behavior
     std::ptr::write(
     //~^ ERROR assigning to `&T` is undefined behavior

--- a/tests/ui/lint/reference_casting.rs
+++ b/tests/ui/lint/reference_casting.rs
@@ -130,6 +130,7 @@ unsafe fn assign_to_ref() {
     }
 }
 
+const RAW_PTR: *mut u8 = 1 as *mut u8;
 unsafe fn no_warn() {
     let num = &3i32;
     let mut_num = &mut 3i32;
@@ -144,6 +145,9 @@ unsafe fn no_warn() {
     let mut value = 3;
     let value: *const i32 = &mut value;
     *(value as *const i16 as *mut i16) = 42;
+    *RAW_PTR = 42; // RAW_PTR is defined outside the function body,
+                   // make sure we don't ICE on it when trying to
+                   // determine if we should lint on it or not.
 
     fn safe_as_mut<T>(x: &std::cell::UnsafeCell<T>) -> &mut T {
         unsafe { &mut *std::cell::UnsafeCell::raw_get(x as *const _ as *const _) }

--- a/tests/ui/lint/reference_casting.stderr
+++ b/tests/ui/lint/reference_casting.stderr
@@ -82,6 +82,14 @@ LL |     let _num = &mut *std::mem::transmute::<_, *mut i32>(num);
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
   --> $DIR/reference_casting.rs:39:16
    |
+LL |     let _num = &mut *(std::mem::transmute::<_, *mut i32>(num) as *mut i32);
+   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
+
+error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
+  --> $DIR/reference_casting.rs:41:16
+   |
 LL |       let _num = &mut *std::cell::UnsafeCell::raw_get(
    |  ________________^
 LL | |
@@ -92,7 +100,7 @@ LL | |     );
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:45:16
+  --> $DIR/reference_casting.rs:47:16
    |
 LL |     let deferred = num as *const i32 as *mut i32;
    |                    ----------------------------- casting happend here
@@ -102,7 +110,7 @@ LL |     let _num = &mut *deferred;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:48:16
+  --> $DIR/reference_casting.rs:50:16
    |
 LL |     let deferred = (std::ptr::from_ref(num) as *const i32 as *const i32).cast_mut() as *mut i32;
    |                    ---------------------------------------------------------------------------- casting happend here
@@ -112,7 +120,7 @@ LL |     let _num = &mut *deferred;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:51:16
+  --> $DIR/reference_casting.rs:53:16
    |
 LL |     let deferred = (std::ptr::from_ref(num) as *const i32 as *const i32).cast_mut() as *mut i32;
    |                    ---------------------------------------------------------------------------- casting happend here
@@ -123,7 +131,7 @@ LL |     let _num = &mut *deferred_rebind;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:53:16
+  --> $DIR/reference_casting.rs:55:16
    |
 LL |     let _num = &mut *(num as *const _ as usize as *mut i32);
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -131,7 +139,15 @@ LL |     let _num = &mut *(num as *const _ as usize as *mut i32);
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:60:16
+  --> $DIR/reference_casting.rs:57:16
+   |
+LL |     let _num = &mut *(std::mem::transmute::<_, *mut _>(num as *const i32) as *mut i32);
+   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
+
+error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
+  --> $DIR/reference_casting.rs:64:16
    |
 LL |     let num = NUM as *const i32 as *mut i32;
    |               ----------------------------- casting happend here
@@ -142,7 +158,7 @@ LL |     let _num = &mut *num;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:64:9
+  --> $DIR/reference_casting.rs:68:9
    |
 LL |         &mut *((this as *const _) as *mut _)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -150,7 +166,7 @@ LL |         &mut *((this as *const _) as *mut _)
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:69:18
+  --> $DIR/reference_casting.rs:73:18
    |
 LL |         unsafe { &mut *std::cell::UnsafeCell::raw_get(x as *const _ as *const _) }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -158,7 +174,7 @@ LL |         unsafe { &mut *std::cell::UnsafeCell::raw_get(x as *const _ as *con
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: casting `&T` to `&mut T` is undefined behavior, even if the reference is unused, consider instead using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:74:18
+  --> $DIR/reference_casting.rs:78:18
    |
 LL |         unsafe { &mut *std::cell::UnsafeCell::raw_get(x as *const _ as *const _) }
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -166,7 +182,7 @@ LL |         unsafe { &mut *std::cell::UnsafeCell::raw_get(x as *const _ as *con
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:84:5
+  --> $DIR/reference_casting.rs:88:5
    |
 LL |     *(a as *const _ as *mut _) = String::from("Replaced");
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -174,7 +190,7 @@ LL |     *(a as *const _ as *mut _) = String::from("Replaced");
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:86:5
+  --> $DIR/reference_casting.rs:90:5
    |
 LL |     *(a as *const _ as *mut String) += " world";
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -182,7 +198,7 @@ LL |     *(a as *const _ as *mut String) += " world";
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:88:5
+  --> $DIR/reference_casting.rs:92:5
    |
 LL |     *std::ptr::from_ref(num).cast_mut() += 1;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -190,7 +206,7 @@ LL |     *std::ptr::from_ref(num).cast_mut() += 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:90:5
+  --> $DIR/reference_casting.rs:94:5
    |
 LL |     *std::ptr::from_ref({ num }).cast_mut() += 1;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -198,7 +214,7 @@ LL |     *std::ptr::from_ref({ num }).cast_mut() += 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:92:5
+  --> $DIR/reference_casting.rs:96:5
    |
 LL |     *{ std::ptr::from_ref(num) }.cast_mut() += 1;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -206,7 +222,7 @@ LL |     *{ std::ptr::from_ref(num) }.cast_mut() += 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:94:5
+  --> $DIR/reference_casting.rs:98:5
    |
 LL |     *(std::ptr::from_ref({ num }) as *mut i32) += 1;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -214,7 +230,7 @@ LL |     *(std::ptr::from_ref({ num }) as *mut i32) += 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:96:5
+  --> $DIR/reference_casting.rs:100:5
    |
 LL |     *std::mem::transmute::<_, *mut i32>(num) += 1;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -222,7 +238,15 @@ LL |     *std::mem::transmute::<_, *mut i32>(num) += 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:98:5
+  --> $DIR/reference_casting.rs:102:5
+   |
+LL |     *(std::mem::transmute::<_, *mut i32>(num) as *mut i32) += 1;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |
+   = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
+
+error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
+  --> $DIR/reference_casting.rs:104:5
    |
 LL | /     std::ptr::write(
 LL | |
@@ -234,7 +258,7 @@ LL | |     );
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:105:5
+  --> $DIR/reference_casting.rs:111:5
    |
 LL |     let value = num as *const i32 as *mut i32;
    |                 ----------------------------- casting happend here
@@ -244,7 +268,7 @@ LL |     *value = 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:108:5
+  --> $DIR/reference_casting.rs:114:5
    |
 LL |     let value = num as *const i32 as *mut i32;
    |                 ----------------------------- casting happend here
@@ -255,7 +279,7 @@ LL |     *value_rebind = 1;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:110:5
+  --> $DIR/reference_casting.rs:116:5
    |
 LL |     *(num as *const i32).cast::<i32>().cast_mut() = 2;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -263,7 +287,7 @@ LL |     *(num as *const i32).cast::<i32>().cast_mut() = 2;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:112:5
+  --> $DIR/reference_casting.rs:118:5
    |
 LL |     *(num as *const _ as usize as *mut i32) = 2;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
@@ -271,7 +295,7 @@ LL |     *(num as *const _ as usize as *mut i32) = 2;
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:114:5
+  --> $DIR/reference_casting.rs:120:5
    |
 LL |     let value = num as *const i32 as *mut i32;
    |                 ----------------------------- casting happend here
@@ -282,7 +306,7 @@ LL |     std::ptr::write(value, 2);
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:116:5
+  --> $DIR/reference_casting.rs:122:5
    |
 LL |     let value = num as *const i32 as *mut i32;
    |                 ----------------------------- casting happend here
@@ -293,7 +317,7 @@ LL |     std::ptr::write_unaligned(value, 2);
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:118:5
+  --> $DIR/reference_casting.rs:124:5
    |
 LL |     let value = num as *const i32 as *mut i32;
    |                 ----------------------------- casting happend here
@@ -304,12 +328,12 @@ LL |     std::ptr::write_volatile(value, 2);
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
 error: assigning to `&T` is undefined behavior, consider using an `UnsafeCell`
-  --> $DIR/reference_casting.rs:122:9
+  --> $DIR/reference_casting.rs:128:9
    |
 LL |         *(this as *const _ as *mut _) = a;
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: for more information, visit <https://doc.rust-lang.org/book/ch15-05-interior-mutability.html>
 
-error: aborting due to 35 previous errors
+error: aborting due to 38 previous errors
 

